subsys/bluetooth/audio/vocs.c

```diff
@@ -0,0 +1,482 @@
+	}
+
+	if (!instances_prepared) {
+		prepare_vocs_instances();
+		instances_prepared = true;
+	}
+
+	CHECKIF(vocs->srv.initialized) {
+		BT_DBG("Already initialized VOCS instance");
+		return -EALREADY;
+	}
+
+	CHECKIF(init->offset > BT_VOCS_MAX_OFFSET || init->offset < BT_VOCS_MIN_OFFSET) {
+		BT_DBG("Invalid offset %d", init->offset);
+		return -EINVAL;
+	}
+
+	vocs->srv.location = init->location;
+	vocs->srv.state.offset = init->offset;
+
+	if (init->output_desc) {
+		strncpy(vocs->srv.output_desc, init->output_desc,
+			sizeof(vocs->srv.output_desc) - 1);
+		/* strncpy may not always null-terminate */
+		vocs->srv.output_desc[sizeof(vocs->srv.output_desc) - 1] = '\0';
+		if (IS_ENABLED(CONFIG_BT_DEBUG_VOCS) &&
+		    strcmp(vocs->srv.output_desc, init->output_desc)) {
+			BT_DBG("Output desc clipped to %s", log_strdup(vocs->srv.output_desc));
+		}
+	}
+
+	/* Iterate over the attributes in VOCS (starting from i = 1 to skip the service declaration)
+	 * to find the BT_UUID_VOCS_DESCRIPTION or BT_UUID_VOCS_LOCATION and update the
+	 * characteristic value (at [i]), update with the write permission and callback, and
+	 * also update the characteristic declaration (always found at [i - 1]) with the
+	 * BT_GATT_CHRC_WRITE_WITHOUT_RESP property.
+	 */
+	for (int i = 1; i < vocs->srv.service_p->attr_count; i++) {
+		attr = &vocs->srv.service_p->attrs[i];
+
+		if (init->location_writable && !bt_uuid_cmp(attr->uuid, BT_UUID_VOCS_LOCATION)) {
+			/* Update attr and chrc to be writable */
+			chrc = vocs->srv.service_p->attrs[i - 1].user_data;
+			attr->write = write_location;
+			attr->perm |= BT_GATT_PERM_WRITE_ENCRYPT;
+			chrc->properties |= BT_GATT_CHRC_WRITE_WITHOUT_RESP;
+		} else if (init->desc_writable &&
+			   !bt_uuid_cmp(attr->uuid, BT_UUID_VOCS_DESCRIPTION)) {
+			/* Update attr and chrc to be writable */
+			chrc = vocs->srv.service_p->attrs[i - 1].user_data;
+			attr->write = write_output_desc;
+			attr->perm |= BT_GATT_PERM_WRITE_ENCRYPT;
+			chrc->properties |= BT_GATT_CHRC_WRITE_WITHOUT_RESP;
+		}
+	}
+
+	err = bt_gatt_service_register(vocs->srv.service_p);
+	if (err) {
+		BT_DBG("Could not register VOCS service");
+		return err;
+	}
+
+	vocs->srv.initialized = true;
+	return 0;
+}
+#endif /* CONFIG_BT_VOCS */
+
+#if defined(CONFIG_BT_VOCS) || defined(CONFIG_BT_VOCS_CLIENT)
+
+int bt_vocs_state_get(struct bt_conn *conn, struct bt_vocs *inst)
+{
+	CHECKIF(!inst) {
+		BT_DBG("Null VOCS pointer");
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_VOCS_CLIENT) && conn) {
+		return bt_vocs_client_state_get(conn, inst);
+	} else if (IS_ENABLED(CONFIG_BT_VOCS) && !conn) {
+		if (inst->srv.cb && inst->srv.cb->state) {
+			inst->srv.cb->state(NULL, inst, 0, inst->srv.state.offset);
+		}
+		return 0;
+	}
+
+	return -EOPNOTSUPP;
+}
+
+int bt_vocs_location_get(struct bt_conn *conn, struct bt_vocs *inst)
+{
+	CHECKIF(!inst) {
+		BT_DBG("Null VOCS pointer");
+		return -EINVAL;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_VOCS_CLIENT) && conn) {
```

Comment:
Can the application be both client and server? And would that cause a problem when you call location get and you get the client always?

Making this comment once, but it applies to all with the same pattern.

Reply:
Yes, the application can be both client and server. As per the API, setting `conn = NULL` will get the local server value, while `conn != NULL` will attempt to read the value from a remote server as the client. If this is unclear, we need to add more information in the documentation.

Reply:
I think that is a bit unclear, my suggestion would be to have two functions.
Local values for the server should be `_get`/`_set`, and the clients `_read`/`_write` to show the GATT operation.